CLOUDP-93933: Improve CONTRIBUTING docs (#730)
mongodb/mongocli: CONTRIBUTING.md

Please create a [GitHub issue](https://github.com/mongodb/mongocli/issues/new?assignees=&labels=&template=bug_report.md) describing the kind of problem you're facing
with as much detail as possible, including things like operating system or anything else may be relevant to the issue.

## Submitting Patches

The MongoDB CLI project welcomes all contributors and contributions regardless of skill or experience level. 

> If you are interested in helping with the project, we will help you with your contribution.

Reviewer: I really like this statement 💯

Reviewer 2: how about directing them to the guidelines ? 
```suggestion
If you are interested in helping with the project, please follow our [guidelines](#code-contribution-guidelines).
```

Reviewer 3: +1

Author: @colm-quinn is the +1 for andrea or for filipe, trying to decide if to keep it as it or not and you are the deciding vote

Reviewer 3: 🥁 it's for Filipe :)

Reviewer 2: it's for both @gssbzn, @colm-quinn both likes the comment and the suggestion 😆

### Code Contribution Guidelines

> We want to create the best possible product for our users and the best contribution experience for our developers,

Reviewer 4: This sentence and the next can be connected a bit more seamlessly. One way would be to remove `We want ` (Start the sentence with `To create...`).

we have a set of guidelines to ensure that all contributions are acceptable.

To make the contribution process as seamless as possible, we ask for the following:

* Go ahead and fork the project and make your changes.  

>   We encourage pull requests to allow for review and discussion of code changes.

Reviewer 4: If pull requests is a requirement for code changes to become part of the repo, maybe this part can be worded a bit differently. For example:
```
* Fork the repository to work on your changes. Note that code contributions are accepted through pull requests, to encourage discussion and allow for a smooth review experience. 

```

* When you’re ready to create a pull request, be sure to:
  * Sign the [CLA](https://www.mongodb.com/legal/contributor-agreement).

>   * Have test cases for the new code. If you have questions about how to do this, please ask in your pull request.

Reviewer 3: [nit] I'd link to `Generating Mocks` at this point to help users write well-structured tests.

  * Run `make fmt`.

>   * Add documentation if you are adding new features or changing functionality.  

Reviewer 3: [nit] Link to the `Maintainer's Guide` section here or explicitly call out that docs needs to be tagged (it's easy to forget and a prompt to the user is very helpful)

Also, do we have a note in the new PR template with that note, so again people see it everytime they make a PR?

Author: I'm thinking about this, I don't want to link to maintainer's guide to try and not overwhelm people here we have a field to signal you added docs but may be lost, trying to consider if we can use a github action to add labels or something as a CODEOWNERS may be hard to add for this point

Reviewer 3: No problem - it's a nit, just want to help people be able to figure out what to do as easily as possible.

Author: I looked at https://github.com/actions/labeler but I can only filter by files and this may cover flags as we have then centralize but for commands it may be a bit more complex so I'll let this as tbd

>   * Ensure that `make check` succeeds. [GitHub Actions](https://github.com/mongodb/mongocli/actions).

Reviewer 4: Slight correction, `be sure to: ensure that` ->  `be sure to: confirm that`

We provide a git pre-commit hook to format and check the code, to install it run `make link-git-hooks`.

#### Generating Mocks

We use [mockgen](https://github.com/golang/mock) to handle mocking in our unit tests.
If you need a new mock please update or add the `//go:generate` instruction to the appropriate file.

### Adding a New Command

`mongocli` uses [Cobra](https://github.com/spf13/cobra) as a framework for defining commands,
in addition to this we have defined a basic structure that should be followed.
For a `mongocli scope newCommand` command a file `internal/cli/scope/new_command.go` should implement: 
- A `ScopeNewCommandOpts` struct which handles the different options for the command.
- At least a `func (opts *ScopeNewCommandOpts) Run() error` function with the main command logic.
- A `func ScopeNewCommandBuilder() *cobra.Command` function to put together the expected cobra definition along with the `ScopeNewCommandOpts` logic.

Commands try to follow a [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) approach to match the APIs, 

> this means arguments to the command try to match path and query params where the last path param will be a required argument of the command and teh rest will be handled via flag options.

Reviewer: ```suggestion
this means arguments to the command try to match path and query params where the last path param will be a required argument of the command and the rest will be handled via flag options.
```

For commands that create or modify complex data structure the use of configuration files is preferred over flag options 

> 

Reviewer 4: A slight rephrasing and punctuation suggestion:

Commands follow a [RESTful](https://en.wikipedia.org/wiki/Representational_state_transfer) approach to match the APIs, whenever possible. For that reason, command arguments tend to match the path and query params of the APIs, with the last param being a required argument and the rest handled via flag options. For commands that create or modify complex data structures, the use of configuration files is preferred over flag options.

Author: thanks, I really wanted comments on this paragraph and this is way better
